language/move-model/src/ty.rs

```diff
@@ -830,13 +792,34 @@ impl TypeUnificationAdapter {
     fn new<'a, I>(
         lhs_types: I,
         rhs_types: I,
-        treat_type_param_as_var: bool,
-        treat_type_local_as_var: bool,
+        treat_lhs_type_param_as_var: bool,
+        treat_rhs_type_param_as_var: bool,
     ) -> Self
     where
         I: Iterator<Item = &'a Type> + Clone,
     {
+        assert!(
+            treat_rhs_type_param_as_var || treat_lhs_type_param_as_var,
+            "At least one side of the unification must be treated as variable"
+        );
+        // Compute the number of type parameters for each side.
+        let mut lhs_type_param_count = 0;
+        let mut rhs_type_param_count = 0;
+        let count_type_param = |t: &Type, current: &mut u16| {
+            if let Type::TypeParameter(idx) = t {
+                *current = (*current).max(*idx + 1);
+            }
+        };
+        for ty in lhs_types.clone() {
+            ty.visit(&mut |t| count_type_param(t, &mut lhs_type_param_count));
+        }
+        for ty in rhs_types.clone() {
+            ty.visit(&mut |t| count_type_param(t, &mut rhs_type_param_count));
+        }
+
         // find the starting index for new type vars
+        // TODO: perhaps we should not allow this function to be called with type vars
+        //   in lhs or rhs?
```

Comment:
Yes, I agree. Maybe we should ban it.

Reply:
We should also add a note somewhere saying that the `TypeUnificationAdapter` can only be called when the `TypeParameter` for LHS and RHS comes from different origins.

Reply:
Done